feat: Flat and differential runtime config files

jakmeier opened the pull request:
Replace the JSON runtime config file per version with a base
`parameters.txt` file plus differential files per version.
The `RuntimeConfig` remains untouched. But it is now loaded from the
flat list of parameters, that converts it to JSON structured data following
the exact layout we use in code, that is then converted to an actual struct
using serde derived code.

The new TXT files are written in a custom format for key-values separated
by a column. Keys must match a defined parameter and values are strings 
or integers. Differential files include old and new values with a `->` between
them.

The direct benefit of this is better readability for config changes and
an easier and less error-prone process of doing such changes.

Indirectly, this could be the first step to further clean up code
around gas costs.
(1) the gas profile could use these parameters to track costs instead
of its own `enum Cost`.
(2) The params-estimator can link estimations to params more easily.

Test plan
---------
The usage of `RuntimeConfigStore` has not changed, only how it is
created. Therefore, tests can ensure all stored `RuntimeConfig`
versions are exactly the same as before. And this should be sufficient
to confirm this change has no side-effects.

To implement this, we keep the old JSON files in the `legacy_configs`
directory. The test `test_old_and_new_runtime_config_format_match` then
iterates through all existing versions and compares the serde
deserialization with configs loaded from TXT files. This is all we need to
verify correctness of the parameters loaded through the new files.

The new TXT file format also comes with custom parsing logic. Therefore,
 a few unit tests to make sure it behaves the way we think in all
circumstances.

nagisa commented on core/primitives/res/runtime_configs/parameters.txt:
```diff
@@ -0,0 +1,152 @@
+max_length_storage_key: 4194304
+max_length_storage_value: 4194304
+max_promises_per_function_call_action: 1024
+max_number_input_data_dependencies: 128
```
Is there a design document around the configuration format? I feel like there may be alternatives that could possibly serve our use-cases slightly better. For example maybe a reverse object notation such as this

`wasm_config.limit_config.max_number_input_data_dependencies`?

would allow us to continue using serde infrastructure (though we'd need to implement a serde deserializer) or perhaps we could use something like `toml` or `json` and merge the files together.

In particular I had non-terrible experience with https://docs.rs/config/latest/config/ and TOML.

matklad replied:
Yeah, some bounded bikeshed would help here!

>would allow us to continue using serde infrastructure 

I wounder if we can go in the opposite direction and actually *remove* a bunch of serializes after this PR? Probably not, because, IIRC, we expose config-as-json over RPC API. If that's indeed the case, yeah, there's a strong reason to try to take advantage of serde. If we want to keep the current format (which I quite like, especialy if we add thousand separators), than we can potentially go `.txt -> serde_json::Value -> RuntimeConfig`

jakmeier replied:
@nagisa I have written a quick design-document in https://github.com/near/nearcore/issues/6710. I should have created an issue before submitting this PR, thanks for pointing it out! But I am also trying to keep things simple. Instead of putting too much time into finding the perfect solution, I would rather spend more time on writing tests that the chosen solution has no observable side-effects.

That said, I appreciate all inputs! As outlined in the other issue, keeping the file structured in any way was something I explicitly tried to avoid. The motivation that I even started to work on this was to avoid a lot of bloated code in PR https://github.com/near/nearcore/pull/6657 where @matklad suggested to first clean up the mess with parameters/profiled costs/runtime config before linking them to estimations. If you see a way to solve the problems I described in the other issue without fully-flattening, please let me know.

But I am open to use any sort of standardised format + serde. Although, I don't really see the value of it if we go the fully-flat route.

nagisa commented on core/primitives-core/src/parameter.rs:
```diff
@@ -0,0 +1,202 @@
+    key: Parameter,
+) -> Option<F> {
+    let key_str: &'static str = key.into();
+    Some(params.get(&key)?.parse().unwrap_or_else(|_err| {
```
Hm, I'm not sure if `FromStr` is the best tool to use here. Notably there is no implementation of `FromStr` for things like `Option<_>` and so there is no way to reset a parameter to `None` if the parameter was already set to `Some` before.

jakmeier replied:
I think I now got rid of all uses of `FromStr`. Parsing to types is mostly left to `serde`. Although, the semantics of empty, missing, or null values is still a bit awkward

nagisa commented on core/primitives-core/src/parameter.rs:
```diff
@@ -0,0 +1,202 @@
+}
+
+pub fn read_optional_parameter<F: std::str::FromStr>(
+    params: &BTreeMap<Parameter, String>,
```
I suspect a newtype over the `BTreeMap` would make a lot of sense API design wise.

jakmeier replied:
Good point. In my latest commit I defined this:
```
pub(crate) struct ParameterTable {
    params: BTreeMap<Parameter, String>,
}
```

nagisa commented on core/primitives-core/src/parameter.rs:
```diff
@@ -0,0 +1,202 @@
+            let mut iter = trimmed.split(":");
+            let key_str = iter.next().unwrap().trim();
+            let typed_key = Parameter::from_str(key_str)
+                .unwrap_or_else(|_err| panic!("Unexpected parameter `{key_str}`."));
```
I would like to see these panics to become regular errors here. There's a lot of value in giving the caller of this function the choice to panic or to handle the error in some other way, I feel. For example it'd be pretty difficult to write unit tests for this functionality (which is something that should be added before landing this PR)

jakmeier replied:
That makes sense. What error type would you use here? Something simple as `std::err::Result<_,String>`? Or would you define a new error type in this case?

It's one of those things where I have seen too many solutions over the years but never dug into the details enough to understand which solution is generally accepted to fit which situation.

nagisa replied:
A `thiserror::Error` error type would be most appropriate.

jakmeier replied:
Thanks for the advice. Now I propagate a `thiserror::Error` up to `RuntimeConfigStore::new()`, where it still panics. But it is now unit-testable. And when it panics, the error message has context about which file caused the error, so definitely an improvement.

nagisa commented on core/primitives/res/runtime_configs/42.txt:
```diff
@@ -0,0 +1 @@
+storage_amount_per_byte: 10000000000000000000
```
It would be great to have a test that would verify that or all of these configs end up the same as the data in legacy configs.

matklad replied:
I think that's `test_runtime_config_data`? It would be great if we can re-use that test without actually continuing to carry JSONs. I think hashes today are actually hashes of JSON, but we can probably use StableHash of the RuntimeConfig struct itself? If we *first* change the test, and than add parameters, we'll check that hashes are the same.

matklad commented on core/primitives-core/src/config.rs:
```diff
@@ -419,6 +480,118 @@ pub fn default_read_cached_trie_node() -> Gas {
+            alt_bn128_g1_sum_element: read_parameter(params, Parameter::WasmAltBn128G1SumElement),
+        }
+    }
+
```
let's maybe move these bulky statements to the `parameter` module?

jakmeier replied:
My latest commit moves all new constructors to `parameter_table` but left other constructors where they are. Not sure if this is what you had in mind.

matklad commented on core/primitives-core/src/parameter.rs:
```diff
@@ -0,0 +1,202 @@
+use std::{
+    collections::BTreeMap,
+    io::{BufRead, BufReader},
+    str::FromStr,
+};
```
We don't group imports like that:

https://github.com/near/nearcore/blob/master/docs/style.md#import-granularity

jakmeier replied:
fixed in latest commit

matklad commented on core/primitives-core/src/parameter.rs:
```diff
@@ -0,0 +1,202 @@
+}
+
+pub fn load_parameters_from_txt(arg: &[u8]) -> BTreeMap<Parameter, String> {
+    let parameters = BufReader::new(arg).lines().filter_map(|line| {
```
Or, perhaps even better, push the invariant to the caller and use `&str` as an argument (changing the caller to use `include_str` and pushing utf-8 validation to rustc, for the ultimate speed :-) )

matklad commented on core/primitives-core/src/parameter.rs:
```diff
@@ -0,0 +1,202 @@
+        if trimmed.starts_with("#") || trimmed.is_empty() {
+            None
+        } else {
+            let mut iter = trimmed.split(":");
```
`split_once` is probably more convenient API.

matklad commented on core/primitives-core/src/parameter.rs:
```diff
@@ -0,0 +1,202 @@
+        } else {
+            let mut iter = trimmed.split(":");
+            let key_str = iter.next().unwrap().trim();
+            let typed_key = Parameter::from_str(key_str)
```
```suggestion
            let typed_key: Parameter = key_str.parse()
```

as a rule of thumb, never call from_str, always call parse

matklad commented on core/primitives-core/src/runtime/fees.rs:
```diff
@@ -136,6 +157,58 @@ pub struct StorageUsageConfig {
+                read_parameter(params, Parameter::PessimisticGasPriceInflationDenominator),
+            ),
+        }
+    }
```
Yeah, I think this also should be in the parameters module.

matklad commented on core/primitives/res/runtime_configs/49.txt:
```diff
@@ -0,0 +1,2 @@
+wasm_regular_op_cost: 822756
```
I wonder if it makes to spell this as

```
wasm_regular_op_cost: 822756 -> 2207874
```

Extra check, and somewhat better readability. 

Not sure if its worth the extra impl complexity

jakmeier replied:
I like it a lot! In the current format, implementing this would be simple enough, I think. But I will wait to see if we might want to go to a different format instead.

matklad commented on core/primitives/res/runtime_configs/parameters.txt:
```diff
@@ -0,0 +1,152 @@
+action_receipt_creation_send_sir: 108059500000
+action_receipt_creation_send_not_sir: 108059500000
+action_receipt_creation_execution: 108059500000
+data_receipt_creation_base_send_sir: 4697339419375
```
Can we format numbers with thousand separators? `4_697_339_419_375`

jakmeier replied:
Done

matklad commented on nearcore/src/runtime/mod.rs:
```diff
@@ -264,7 +265,7 @@ impl NightshadeRuntime {
             "testnet" => {
                 let genesis_runtime_config =
-                    serde_json::from_slice(INITIAL_TESTNET_CONFIG).unwrap();
+                    RuntimeConfig::from_parameters_txt(INITIAL_TESTNET_CONFIG);
```
FWIW, feels like an abstraction leakage. runtime shouldn't care if we store config in json or txt. I'd maybe add `fn RuntimeConfig::initial_testnet_config` instead?

matklad commented on core/primitives-core/src/parameter.rs:
```diff
@@ -0,0 +1,202 @@
+    MaxLocalsPerContract,
+}
+
+pub fn load_parameters_from_txt(arg: &[u8]) -> BTreeMap<Parameter, String> {
```
```suggestion
pub(crate) fn load_parameters_from_txt(arg: &[u8]) -> BTreeMap<Parameter, String> {
```

I think it makes sense to avoid `pub` as much as possible. Cross-crate deps are the only deps that matter, so they should stand out.

jakmeier replied:
Hm, I defined this in `primitives-core` crate, where as all callers are in `primitives` crate. So your suggestion doesn't work out of the box. But is also seems wrong to have code define on crate A and only call it in crate B.

However, the type `Parameter` must be in `primitives-core` because it should be the new source-of-truth replacing some types in `core/primitives-core/src/profile.rs`. And `primitives-core` is not allowed to depend on other crates, according to our architecture.md.

So, in my latest commit, I moved all the logic to construct over to `primitives`. To still have the constructor syntax like `VMLimitConfig::from_parameters(params)`, I defined an extension trait `FromParameterTable`. (I can also rewrite to `params.vm_limit_config()` if you think that's better.)

This refactor sort of works but also brings new problems. I previously liked that each config struct had all constructors (for testing and for production) right next to each other, whereas now they are defined in different crates. And there are other things, like `wasmer2_stack_limit_default` that used to be private but now turned public in the refactor.

Not sure if it is an overall improvement or not.

jakmeier commented:
I wrote up the motivation behind this in a separate issue: https://github.com/near/nearcore/issues/6710

nagisa commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -242,3 +243,209 @@ impl FromParameterTable for RuntimeFeesConfig {
+    use std::collections::BTreeMap;
+
+    #[track_caller]
+    fn check_parameter_table(base_config: &str, diffs: &[&str], expected: &ParameterTable) {
```
This would probably be a little more straightforward to use with a signature like this:

```suggestion
    fn check_parameter_table(base_config: &str, diffs: &[&str], expected: impl IntoIterator<Item = (Parameter, &str)>) {
```

and then created the btreemap or whatever other structures to compare with internally.

Then the callers could look something like this:

```rust
        check_parameter_table(
            base_0,
            &[],
            [
                (Parameter::RegistrarAccountId, "registrar"),
                (Parameter::MinAllowedTopLevelAccountLength, "32"),
                (Parameter::StorageAmountPerByte, "100000000000000000000"),
                (Parameter::StorageNumBytesAccount, "100"),
                (Parameter::StorageNumExtraBytesRecord, "40"),
            ],
        );
```

jakmeier replied:
Yes, much cleaner! Done as you suggested

nagisa commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -22,6 +22,7 @@ pub(crate) struct ParameterTable {
 /// Error returned by ParameterTable::from_txt() that parses a runtime
 /// configuration TXT file.
 #[derive(Error, Debug)]
+#[cfg_attr(test, derive(PartialEq))]
```
I think it is okay if you derived `PartialEq` unconditionally here. Another useful derive could be `Clone`.

jakmeier replied:
Adding in the strum errors as source counteracts that, sadly. I ended up removing `PartialEq` entirely and instead use `assert_machtes!`.

jakmeier commented:
Okay, this is ready for review. You probably want to look through all files again but here is the summary of my changes.
- The `ParameterTable` now stores `serde_json::Value` instead of `String`.
- To convert from `ParameterTable` to `RuntimeConfig`, I go through JSON and `serde`-derived code. While creating the JSON, the WASM external costs and the limit configurations are constructed by iterating over the corresponding parameters. All other parameters are still placed manually.
- `Fee`s, that is all action + receipt costs that have a split between `send_sir`/`send_not_sir`/`execution` are listed in `enum FeeParameter` which makes constructing the JSON a bit easier.
- Differential and base files now use slightly different format. (base: `key: value` and diff: `key: new -> old`)
- Addressed all comments from reviews so far

I think the big-picture questions left for me are:
1. Are we all okay with using a custom format? Or should we use a standard format. I could live with both and the work to go from what we have now to a standard format is moderate.
2. Does the TXT -> JSON -> Rust struct strategy make sense the way I implemented it? It is roughly what I understood based on review comments and a quick discussion with matklad yesterday. I think in principle it does make a lot of sense to use `serde` for the final type fitting. But there are a couple of rough corners when actually implementing it. Ideas for improvement are welcome.
3. Have I forgotten things that could benefit from tests?

jakmeier commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -242,3 +243,209 @@ impl FromParameterTable for RuntimeFeesConfig {
+"#;
+
+        // Check empty input
+        check_parameter_table("", &[], &ParameterTable { params: BTreeMap::new() });
```
Yeah no problem, split them up in separate tests. And it already helped me while I was changing the format for diffs

jakmeier commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -18,24 +19,40 @@ pub(crate) struct ParameterTable {
+#[derive(Error, Debug)]
+pub(crate) enum InvalidConfigError {
+    #[error("Unknown parameter `{0}`")]
+    UnknownParameter(String),
```
I see, makes sense to keep the link to the source error. Done

jakmeier commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -18,24 +19,40 @@ pub(crate) struct ParameterTable {
+            })
+            .map(|trimmed| {
+                let (key, value) =
+                    trimmed.split_once(":").ok_or(InvalidConfigError::NoSeparator)?;
```
Right. The error message now includes both line number and the text.

nagisa commented on core/primitives/res/README.md:
```diff
@@ -8,10 +8,16 @@ All parameter value to configure the runtime are defined in `parameters.txt`.
 `core/primitives-core/src/parameter.rs`.
 
+The format of the differential files is slightly different. While the keys are
+the same, values that change use a format lie this: `key: old_value ->
```
```suggestion
Parameters that change are specified like this: `key: old_value -> new_value`.
```

nagisa commented on core/primitives/res/README.md:
```diff
@@ -8,10 +8,16 @@ All parameter value to configure the runtime are defined in `parameters.txt`.
 
+The format of the differential files is slightly different. While the keys are
+the same, values that change use a format lie this: `key: old_value ->
+new_value`. Deleting a parameter value for the new version is done as follows:
+`key: old_value ->`. Finally, inserting is done using the same syntax as in the
```
It isn't super clear to me what “deleting” means here. My intuition says that the merged configuration will be missing this parameter which will in turn use the default value for the parameter, which can potentially be quite different semantically from deleting entirely.

jakmeier replied:
Yes you are right, so I changed the wording to make it clearer.

nagisa commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -137,16 +137,26 @@ impl ParameterTable {
         for param in params {
-            json.insert(param.to_string(), self.get(*param).clone());
+            let mut key: &'static str = param.into();
+            if key.starts_with(remove_prefix) {
+                key = &key[remove_prefix.len()..];
+            }
```
```suggestion
            let key = key.strip_prefix(remove_prefix).unwrap_or(key);
```

nagisa commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -15,32 +20,29 @@ pub(crate) enum InvalidConfigError {
     ValueParseError(#[source] serde_json::Error, String),
-    #[error("Parameter name and value must be separated by a `:`")]
-    NoSeparator,
+    #[error("Parameter name and value must be separated by a `:`. Line {0}: `{1}`")]
```
A convention is that the error Display implementations should be output a non-capitalized sentence. That allows the description to be inserted into another sentence without it looking super out-of-place. Particularly common examples are `error: {}` and `caused by: {}`, but more complex formatting is also a thing, so it is always worth it to take a moment to consider what these strings would look like when incorporated in a broader sentence.

For this particular the wording could be something like

```
expected a `:` separator between name and value of a parameter `{1}` on line {0}
```

or maybe

```
name and value of the parameter `{1}` on line {0} must be separated by a `:`
```

Sorry to nit here again, my bad for not elaborating on this last time I made the suggestions.

jakmeier replied:
Ah now I understand :) Thanks for taking the time to explain. I changed all messages to conform with this idea

nagisa commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -67,8 +69,34 @@ impl ParameterTable {
+                    Some(serde_json::Value::Null) | None => {
+                        self.params.insert(key, after);
+                    }
+                    Some(_) => return Err(InvalidConfigError::OldValueExists(key)),
```
May be worthwhile to preserve what the old value was as well, so that whoever is editing the configuration doesn't need to go spelunking through config files looking for it.

nagisa commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -139,22 +218,27 @@ fn parse_parameter_txt_value(value: &str) -> Result<serde_json::Value, InvalidCo
+        // We do not have "arbitrary_precision" serde feature enabled, thus we
+        // can only store up to `u64::MAX`, which is `18446744073709551615` and
+        // has 20 characters.
+        if raw_number.len() < 20 {
```
Are we happy with the behaviour here when somebody specifies 10^19 which is 20 digits long but is still 2<sup>63.11</sup> and fits into u64? `serde_json` does have an error code for out-of-range numbers internally but sadly it isn't exposed :(

If not, I'm worried that serde_json might not actually be the right tool for us tbh.

nagisa replied:
It probably isn't terribly difficult to implement a custom deserializer just for the values. An example of an implementation that does this is the aforementioned [`config-rs`](https://github.com/mehcode/config-rs/blob/master/src/de.rs) (I encourage to at least skim throug the link here). It has handling for u128 built-in too.

jakmeier replied:
Ok I see, a custom value parser would make sense. Also thanks for the link with the example.

As things stand right now, I don't see a strong motivation to do it right now. As long as we have JSON as intermediary, we have to store numbers as strings anyway and as long as we don't lose precision on the way, I think it doesn't matter too much where the cut-off point is.

nagisa commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -139,22 +218,27 @@ fn parse_parameter_txt_value(value: &str) -> Result<serde_json::Value, InvalidCo
         return Ok(serde_json::Value::Null);
     }
     if value.chars().all(|c| c.is_numeric() || c == '_') {
```
Does json allow all unicode numeric codepoints? May want to check for `is_ascii_digit` (in which case you can also iterate through bytes instead of characters) instead.

jakmeier replied:
I didn't even think about this. But I looked it up, [JSON grammar](https://datatracker.ietf.org/doc/html/rfc7159#section-6) for number values is very strict and has no room for Unicode interpretation. So I changed it as you suggested.

nagisa commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -139,22 +218,27 @@ fn parse_parameter_txt_value(value: &str) -> Result<serde_json::Value, InvalidCo
-                .parse()
-                .map_err(|err| InvalidConfigError::ValueParseError(err, value.to_owned()))?,
-        ))
+        let raw_number = value.chars().filter(|c| c.is_numeric()).collect::<String>();
```
https://doc.rust-lang.org/stable/std/string/struct.String.html#method.retain might work better.

nagisa reviewed (approved):
I think this looks pretty good. I imagine we can change the exact implementation of value deserializer later on if we find that json is not actually suitable for our needs.

That said, please give other reviewers some time to look through this still.

jakmeier commented:
> I think this looks pretty good. I imagine we can change the exact implementation of value deserializer later on if we find that json is not actually suitable for our needs.
> 
> That said, please give other reviewers some time to look through this still.

Thanks! Yes, I am not really in a hurry here. Since some people are taking time off this week, I will probably wait to merge this until next week.

Longarithm commented on core/primitives/res/runtime_configs/49.txt:
```diff
@@ -0,0 +1,2 @@
+wasm_regular_op_cost: 2_207_874-> 822_756
```
```suggestion
wasm_regular_op_cost: 2_207_874 -> 822_756
```
nit

jakmeier replied:
Good catch :)

Longarithm reviewed (approved):
Thank you very much for making config updates more simple and transparent!

matklad commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -0,0 +1,525 @@
+/// A value can be a positive integer or a string, both written without quotes.
+/// Integers can use underlines as separators (for readability).
+fn parse_parameter_txt_value(value: &str) -> Result<serde_json::Value, InvalidConfigError> {
+    if value.is_empty() {
+        return Ok(serde_json::Value::Null);
+    }
```
this check feels like somethig worth pushing to the caller perhaps. Just a gut automatic reaction, maybe this is wrong in context :)

jakmeier replied:
Not sure about this. Most of the parameters are number types and I would rather have it as `null` when they are unspecified. Also, serde will not like that, I think.

matklad commented on core/primitives/src/runtime/config_store.rs:
```diff
@@ -101,25 +114,59 @@ mod tests {
+    }
+
+    #[test]
+    fn test_old_runtime_config_data() {
```
Nice assuarance that we didn't change values in the process of migration! Let's remove it once the PR is merged (that is, I think the only value of this test is to check this specific PR, and, once that it, wec an remove the test without the loss of confidence? )!

matklad reviewed (approved):
Left some minor style comments, but they are optional and can be followed-up!

jakmeier commented on core/primitives/res/runtime_configs/parameters.txt:
```diff
@@ -0,0 +1,151 @@
+# Gas economics config
```
Great idea, I've written something along those lines.

jakmeier commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -0,0 +1,525 @@
+}
+
+impl ParameterTable {
+    pub(crate) fn from_txt(arg: &str) -> Result<ParameterTable, InvalidConfigError> {
```
Yeah, done

jakmeier commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -0,0 +1,525 @@
+                Ok((typed_key, parse_parameter_txt_value(value.trim())?))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        Ok(ParameterTable { params: BTreeMap::from_iter(parameters) })
```
Nice, I always forget how powerful collect is :)

On `params: parameters`, I like your rule of thumb! I didn't really think about it before but I can see how that makes sense. 
=> Changed the field to the full name
